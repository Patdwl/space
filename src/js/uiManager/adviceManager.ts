/* /////////////////////////////////////////////////////////////////////////////

(c) 2016-2020, Theodore Kruczek

advice-module.js manages all recommended actions to the user in a semi-tutorial
manner. It works closely with ui.js
http://keeptrack.space

All rights reserved. No part of this web site may be reproduced, published,
distributed, displayed, performed, copied or stored for public or private
use, without written permission of the author.

No part of this code may be modified or changed or exploited in any way used
for derivative works, or offered for sale, or used to construct any kind of database
or mirrored at any other location without the express written permission of the author.

///////////////////////////////////////////////////////////////////////////// */
import $ from 'jquery';
// eslint-disable-next-line sort-imports
import 'jquery-ui-bundle';
import { keepTrackApi } from '../api/keepTrackApi';
import { AdviceCounter, AdviceList } from '../api/keepTrackTypes';

let isAdviceEnabled = true;
let helpDOM: HTMLDivElement;
let helpCloseDOM: HTMLDivElement;
let helpHeaderDOM: HTMLDivElement;
let helpTextDOM: HTMLDivElement;
let tutIconDOM: HTMLDivElement;
let curFocusDOM: any;
let adviceList: AdviceList;
let adviceCount: AdviceCounter;

export const onReady = (wasAdviceEnabledOverride?: boolean) => {
  // Code Once index.htm is loaded
  let wasAdviceEnabled = null;
  try {
    let pastSelection = localStorage.getItem('isAdviceEnabled');
    if (pastSelection === 'true') wasAdviceEnabled = true;
    if (pastSelection === 'false') wasAdviceEnabled = false;
  } catch (e) {
    console.debug(e);
  }

  if (wasAdviceEnabled || wasAdviceEnabled == null || wasAdviceEnabledOverride) {
    adviceManager.on();

    let adviceDisplayTime = 60 * 1000; // 60s
    adviceManager.adviceArray = [
      adviceList.welcome,
      adviceList.useLegend,
      adviceList.showSensors,
      adviceList.findIss,
      adviceList.missileMenu,
      adviceList.toggleNight,
      adviceList.colorScheme,
      adviceList.customSensors,
      adviceList.countries,
    ];

    // Queue Advice Timers
    for (let i = 0; i < adviceManager.adviceArray.length; i++) {
      setTimeout(adviceManager.adviceArray[i], i * adviceDisplayTime);
    }
  } else {
    adviceManager.off();
  }
  keepTrackApi.methods.adviceReady();
};
export const welcome = (): void => {
  // Only Do this Twice
  if (adviceCount.welcome >= 3) return;
  adviceCount.welcome += 1;

  adviceManager.showAdvice(
    'Welcome',
    'Welcome to KeepTrack.Space! This is the advice system. It will offer ideas on how to use the features on this site. You can enable/disable it at anytime in the top right.',
    $('#tutorial-icon'),
    'top-left'
  );
};
export const findIss = (): void => {
  // Only Do this Twice
  if (adviceCount.findIss >= 3) return;
  adviceCount.findIss += 1;

  adviceManager.showAdvice('Space Station', 'Did you know the international space station is object 25544?', $('#search-holder'), 'top-right');
};
export const showSensors = (): void => {
  // Only Do this Twice Times
  if (adviceCount.showSensors >= 3) return;
  adviceCount.showSensors += 1;

  adviceManager.showAdvice('Sensor List', 'Have you tried looking at some of the ground-based sensors?', $('#menu-sensor-list'), 'bottom-left');
};
export const useLegend = (): void => {
  // Only Do this Twice
  if (adviceCount.useLegend >= 3) return;
  adviceCount.useLegend += 1;

  adviceManager.showAdvice('Filters', "Sometimes it is helpful to filter out satellites you don't want to see right now!", $('#legend-menu'), 'top-right');
};
export const togleNight = (): void => {
  // Only Do this Twice
  if (adviceCount.toggleNight >= 3) return;
  adviceCount.toggleNight += 1;

  adviceManager.showAdvice('Day/Night Toggle', 'Having trouble seeing parts of the earth? You can toggle on/off the night effect.', $('#menu-day-night'), 'bottom');
};
export const missileMenu = (): void => {
  // Only Do this Twice
  if (adviceCount.missileMenu >= 3) return;
  adviceCount.missileMenu += 1;

  adviceManager.showAdvice(
    'Missile Scenarios',
    'Curious how intercontinental ballistic missiles work? Try launching a few from the missile menu.',
    $('#menu-missile'),
    'bottom-right'
  );
};
export const satelliteSelected = (): void => {
  // Only Do this Twice
  if (adviceCount.satelliteView < 1) {
    showSatelliteCameraAdvice();
  } else if (adviceCount.newLaunch < 1) {
    // Only Do this Twice
    showLaunchNominalAdvice();
  } else if (adviceCount.breakup < 1) {
    // Only Do this Twice
    showBreakupAdvice();
  } else if (adviceCount.editSat < 1) {
    // Only Do this Twice
    showEditSatAdvice();
  } else if (adviceCount.satelliteView < 3) {
    showSatelliteCameraAdvice();
  } else if (adviceCount.newLaunch < 3) {
    showLaunchNominalAdvice();
  } else if (adviceCount.breakup < 3) {
    showBreakupAdvice();
  } else if (adviceCount.editSat < 3) {
    // Only Do this Twice
    showEditSatAdvice();
  } else {
    // Intentionally left blank
  }
};
export const colorScheme = (): void => {
  // Only Do this Twice
  if (adviceCount.colorScheme >= 3) return;
  adviceCount.colorScheme += 1;

  adviceManager.showAdvice(
    'Color Schemes',
    'Sometimes it is easier to visualize data by changing the color scheme. You can select from a collection of premade color schemes!',
    $('#menu-color-scheme'),
    'bottom'
  );
};
export const countries = (): void => {
  // Only Do this Twice
  if (adviceCount.countries >= 3) return;
  adviceCount.countries += 1;

  adviceManager.showAdvice(
    'Countries Menu',
    'Did you know that most orbital objects were launched by three countries! Check out the countries menu to view all the satellites launched by a coutnry.',
    $('#menu-countries'),
    'bottom'
  );
};
export const cspocSensors = (): void => {
  // Only Do this Twice
  if (adviceCount.cspocSensors >= 3) return;
  adviceCount.cspocSensors += 1;

  adviceManager.showAdvice(
    'Combined Space Operations Center Sensors',
    'Known by many names (USSPACECOM, 1 SPCS, JSpOC, JSpOC/SSA), the CSpOC coordinates space assets from many countries including United States, United Kingdom, Canada, and Australia.',
    null,
    'right'
  );
};
export const mwSensors = (): void => {
  // Only Do this Twice
  if (adviceCount.mwSensors >= 3) return;
  adviceCount.mwSensors += 1;

  adviceManager.showAdvice(
    'USSTRATCOM Missile Warning Sensors',
    'These sensors primary function is to scan the horizon for potential ICBM launches. Their data is relayed to the Missile Warning Center in Cheyene Mountain, CO.',
    null,
    'right'
  );
};
export const customSensors = (): void => {
  // Only Do this Twice
  if (adviceCount.customSensors >= 3) return;
  adviceCount.customSensors += 1;

  adviceManager.showAdvice(
    'Custom Sensor',
    'Trying to visualize a new sensor or what a current sensor would look like with a different configuration? Try making your own using the custom sensor menu!',
    $('#menu-customSensor'),
    'bottom-left'
  );
};
export const planetariumDisabled = (): void => {
  // Only Do this Twice
  if (adviceCount.planetariumDisabled >= 3) return;
  adviceCount.planetariumDisabled += 1;

  adviceManager.showAdvice(
    'Planetarium View',
    'Using Planetarium View requres a sensor to be selected first. Try selecting a sensor using the Sensor List to find one!',
    $('#menu-sensor-list'),
    'bottom-right'
  );
};
export const satViewDisabled = (): void => {
  // Only Do this Twice
  if (adviceCount.satViewDisabled >= 3) return;
  adviceCount.satViewDisabled += 1;

  adviceManager.showAdvice(
    'Satellite View',
    'Using Satellite View requres a satellite to be selected first. Click on any satellite on the screen and then try using this view again!',
    null,
    'bottom-right'
  );
};
export const mapDIsabled = (): void => {
  // Only Do this Twice
  if (adviceCount.mapDisabled >= 3) return;
  adviceCount.mapDisabled += 1;

  adviceManager.showAdvice(
    'Stereographic Map',
    'Using the Stereographic Map requres a satellite to be selected first. Click on any satellite on the screen and then try loading the map again!',
    null,
    'bottom-right'
  );
};
export const lookanglesDisabled = (): void => {
  // Only Do this Twice
  if (adviceCount.lookanglesDisabled >= 3) return;
  adviceCount.lookanglesDisabled += 1;

  adviceManager.showAdvice(
    'Look Angles',
    'Calculatiung Look Angles requres a satellite and a sensor to be selected first.Pick a sensor, then click any satellite on the screen, and then try calculating Look Angles again!',
    null,
    'bottom-left'
  );
};
export const ssnLookanglesDisabled = (): void => {
  // Only Do this Twice
  if (adviceCount.ssnLookanglesDisabled >= 3) return;
  adviceCount.ssnLookanglesDisabled += 1;

  adviceManager.showAdvice(
    'SSN Lookangles',
    'Calculatiung Look Angles for the whole SSN requres a satellite to be selected first. Click any satellite on the screen, and then try calculating Look Angles for the SSN again!',
    null,
    'bottom-left'
  );
};
export const survFenceDisabled = (): void => {
  // Only Do this Twice
  if (adviceCount.survFenceDisabled >= 3) return;
  adviceCount.survFenceDisabled += 1;

  adviceManager.showAdvice(
    'Show/Hide Surveillance Fence',
    'Displaying the Surveillance Fence requres a sensor to be selected first. Try picking a sensor from the Sensor List and then try again!',
    $('#menu-sensor-list'),
    'bottom-left'
  );
};
export const bubbleDisabled = (): void => {
  // Only Do this Twice
  if (adviceCount.bubbleDisabled >= 3) return;
  adviceCount.bubbleDisabled += 1;

  adviceManager.showAdvice(
    'Show/Hide Field Of View Bubble',
    'Displaying the FOV Bubble requres a sensor to be selected first. Try picking a sensor from the Sensor List and then try again!',
    $('#menu-sensor-list'),
    'bottom-left'
  );
};
export const sensorInfoDisabled = (): void => {
  // Only Do this Twice
  if (adviceCount.sensorInfoDisabled >= 3) return;
  adviceCount.sensorInfoDisabled += 1;

  adviceManager.showAdvice(
    'Sensor Information',
    'Displaying Sensor Information requres a sensor to be selected first. Try picking a sensor from the Sensor List and then try again!',
    $('#menu-sensor-list'),
    'bottom-left'
  );
};
export const editSatDisabled = (): void => {
  // Only Do this Twice
  if (adviceCount.editSatDisabled >= 3) return;
  adviceCount.editSatDisabled += 1;

  adviceManager.showAdvice(
    'Edit Satellite',
    'Editing a Satellite requres a satellite to be selected first. Try picking any satellite from the screen and then try again!',
    null,
    'bottom-right'
  );
};
export const breakupDisabled = (): void => {
  // Only Do this Twice
  if (adviceCount.breakupDisabled >= 3) return;
  adviceCount.breakupDisabled += 1;

  adviceManager.showAdvice(
    'Create Breakup',
    'Creating a Breakup requres a satellite to be selected first. Pick any satellite from the screen and then try again!',
    null,
    'bottom-right'
  );
};
export const satFovDisabled = (): void => {
  // Only Do this Twice
  if (adviceCount.satFovDisabled >= 3) return;
  adviceCount.satFovDisabled += 1;

  adviceManager.showAdvice(
    'Satellite Field of View',
    "Viewing a satellite's field of view requires a satellite to be selected first. Pick any satellite from the screen and then try again!",
    null,
    'bottom'
  );
};
export const sensor = (): void => {
  // Only Do this Twice
  if (adviceCount.sensorFOV < 1) {
    shoFieldOfViewAdvice();
  } else if (adviceCount.sensorSurv < 1) {
    showSurvFenceAdvice();
  } else if (adviceCount.sensorFOV < 3) {
    shoFieldOfViewAdvice();
  } else if (adviceCount.sensorSurv < 3) {
    showSurvFenceAdvice();
  }
};
export const isEnabled = (): boolean => isAdviceEnabled;
export const on = () => {
  try {
    localStorage.setItem('isAdviceEnabled', 'true');
  } catch {
    // Do Nothing
  }
  isAdviceEnabled = true;
  tutIconDOM.addClass('bmenu-item-selected');
};
export const off = () => {
  try {
    localStorage.setItem('isAdviceEnabled', 'false');
  } catch {
    // Do Nothing
  }
  isAdviceEnabled = false;
  helpDOM.hide();
  tutIconDOM.removeClass('bmenu-item-selected');
};
export const showAdvice = (header: string, text: string, focusDOM: HTMLDivElement, setLocation: string) => {
  if (!isAdviceEnabled) return;
  if (typeof setLocation == 'undefined') setLocation = 'bottom-left';
  adviceManager.clearAdvice();
  curFocusDOM = focusDOM;

  switch (setLocation) {
    case 'top-left':
      helpDOM.css({
        left: '1%',
        right: 'auto',
        top: '1%',
        bottom: 'auto',
      });
      break;
    case 'left':
      helpDOM.css({
        left: '1%',
        right: 'auto',
        top: '40%',
        bottom: 'auto',
      });
      break;
    case 'bottom-left':
      helpDOM.css({
        left: '1%',
        right: 'auto',
        top: '60%',
        bottom: 'auto',
      });
      break;
    case 'bottom':
      var leftValue = window.innerWidth / 2 - 175 + 'px';
      helpDOM.css({
        left: leftValue,
        right: 'auto',
        top: '60%',
        bottom: 'auto',
      });
      break;
    case 'top-right':
      helpDOM.css({
        left: 'auto',
        right: '1%',
        top: '1%',
        bottom: 'auto',
      });
      break;
    case 'right':
      helpDOM.css({
        left: 'auto',
        right: '1%',
        top: '40%',
        bottom: 'auto',
      });
      break;
    case 'bottom-right':
      helpDOM.css({
        left: 'auto',
        right: '1%',
        top: '60%',
        bottom: 'auto',
      });
      break;
  }

  helpDOM.show();
  helpHeaderDOM.text(header);
  helpTextDOM.text(text);
  if (typeof focusDOM != 'undefined' && focusDOM != null) {
    focusDOM.effect('shake', { distance: 10 });
    focusDOM.addClass('bmenu-item-help');
    helpHeaderDOM.addClass('help-header-sel');

    helpHeaderDOM.on('click', function () {
      focusDOM.effect('shake', { distance: 10 });
      focusDOM.addClass('bmenu-item-help');
    });
    focusDOM.on('mouseover', function () {
      focusDOM.removeClass('bmenu-item-help');
    });

    focusDOM.on('click', function () {
      focusDOM.removeClass('bmenu-item-help');
      helpDOM.hide();
    });

    helpCloseDOM.on('click', function () {
      helpHeaderDOM.removeClass('help-header-sel');
      focusDOM.removeClass('bmenu-item-help');
      helpHeaderDOM.off();
      if (!focusDOM.is(tutIconDOM)) {
        focusDOM.off();
      }
      helpDOM.hide();
    });
  } else {
    helpCloseDOM.on('click', function () {
      helpDOM.hide();
    });
  }
};
export const clearAdvice = function (): void {
  helpHeaderDOM.removeClass('help-header-sel');
  helpHeaderDOM.off();
  if (typeof curFocusDOM != 'undefined' && curFocusDOM != null) {
    curFocusDOM.removeClass('bmenu-item-help');
    if (!curFocusDOM.is(tutIconDOM)) {
      curFocusDOM.off();
    }
  }
};
export const init = () => {
  helpDOM = $('#help-screen');
  helpDOM.draggable({
    containment: 'window',
    scroll: false,
  });
  helpCloseDOM = $('#help-close');
  helpHeaderDOM = $('#help-header');
  helpTextDOM = $('#help-text');
  tutIconDOM = $('#tutorial-icon');

  // Used on a timer to hint at other ideas
  tutIconDOM.on('click', function () {
    if (isAdviceEnabled) {
      adviceManager.off();
    } else {
      adviceManager.on();
    }
  });

  adviceManager.onReady();
};

export const adviceManager = {
  clearAdvice: clearAdvice,
  isEnabled: isEnabled,
  off: off,
  on: on,
  onReady: onReady,
  showAdvice: showAdvice,
  adviceList: {
    breakupDisabled: breakupDisabled,
    bubbleDisabled: bubbleDisabled,
    colorScheme: colorScheme,
    countries: countries,
    cspocSensors: cspocSensors,
    customSensors: customSensors,
    editSatDisabled: editSatDisabled,
    findIss: findIss,
    lookanglesDisabled: lookanglesDisabled,
    mapDisabled: mapDIsabled,
    missileMenu: missileMenu,
    mwSensors: mwSensors,
    planetariumDisabled: planetariumDisabled,
    satelliteSelected: satelliteSelected,
    satFovDisabled: satFovDisabled,
    satViewDisabled: satViewDisabled,
    sensor: sensor,
    sensorInfoDisabled: sensorInfoDisabled,
    showSensors: showSensors,
    ssnLookanglesDisabled: ssnLookanglesDisabled,
    survFenceDisabled: survFenceDisabled,
    toggleNight: togleNight,
    useLegend: useLegend,
    welcome: welcome,
  },
  adviceCount: {
    breakup: 0,
    breakupDisabled: 0,
    bubbleDisabled: 0,
    colorScheme: 0,
    countries: 0,
    cspocSensors: 0,
    customSensors: 0,
    editSat: 0,
    editSatDisabled: 0,
    findIss: 0,
    lookanglesDisabled: 0,
    mapDisabled: 0,
    missileMenu: 0,
    mwSensors: 0,
    planetariumDisabled: 0,
    satelliteView: 0,
    satFovDisabled: 0,
    satViewDisabled: 0,
    sensorFOV: 0,
    sensorInfoDisabled: 0,
    sensorSurv: 0,
    showSensors: 0,
    ssnLookanglesDisabled: 0,
    survFenceDisabled: 0,
    toggleNight: 0,
    useLegend: 0,
    welcome: 0,
  },
  adviceArray: [] as any[],
};
adviceList = adviceManager.adviceList;
adviceCount = adviceManager.adviceCount;

export { adviceList };

const showEditSatAdvice = () => {
  adviceCount.editSat += 1;

  adviceManager.showAdvice(
    'Edit Satellite',
    'Trying to understand how orbital parameters work? Have you tried editing a satellite to see what the impact is of changing those parameters?',
    $('#menu-editSat'),
    'bottom-right'
  );
};

const showBreakupAdvice = () => {
  adviceCount.breakup += 1;

  adviceManager.showAdvice(
    'Create a Breakup',
    'Curious what this satellite would look like in a 100 pieces? Create a breakup using the breakup menu below! This is also helpful for understanding what a large cubesat launch looks like over time.',
    $('#menu-breakup'),
    'right'
  );
};

const showLaunchNominalAdvice = () => {
  adviceCount.newLaunch += 1;

  adviceManager.showAdvice(
    'Create Launch Nominal',
    'Trying to figure out when a new launch will be in your view? After selecting an ' +
      'object with a similar orbit, use the launch nominal creator menu to position the ' +
      'satellite over the launch site at a time of 0000z. Now you can see details using ' +
      'relative time.',
    $('#menu-newLaunch'),
    'top-right'
  );
};

const showSatelliteCameraAdvice = () => {
  adviceCount.satelliteView += 1;

  adviceManager.showAdvice(
    'Satellite Camera View',
    'Did you know you can change the camera to show what a satellite sees? You have to have a satellite currently selected to use it.',
    $('#menu-satview'),
    'bottom'
  );
};

const showSurvFenceAdvice = () => {
  adviceCount.sensorSurv += 1;
  adviceManager.showAdvice(
    'Surveillance Fence',
    "Most ground-based sensors don't actively look at their entire field of view! They usually scan the horizon. You can see the difference by showing their surveillance fence.",
    $('#menu-surveillance'),
    'bottom-left'
  );
};

const shoFieldOfViewAdvice = () => {
  adviceCount.sensorFOV += 1;
  adviceManager.showAdvice(
    'Field Of View Bubbles',
    'Are you having trouble understanding what a sensor can see? Enable the Field of View Bubble to make it easier to visualize!',
    $('#menu-fov-bubble'),
    'bottom-left'
  );
};
