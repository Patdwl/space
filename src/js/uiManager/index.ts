export * from './input/keyHandler';
export * from './input/uiInput';
export * from './legend/legendMenu';
export * from './menu/menuController';
export * from './uiValidation';
export * from './url/httpsOnly';
