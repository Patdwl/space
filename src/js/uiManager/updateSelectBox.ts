import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { updateInterval } from './uiManager';

export const updateSelectBox = () => {
  const { objectManager, satSet, timeManager } = keepTrackApi.programs;

  // IDEA: Include updates when satellite edited regardless of time.
  // Don't update if no object is selected
  if (objectManager.selectedSat === -1) return;

  const sat = satSet.getSat(objectManager.selectedSat);

  // Don't bring up the update box for static dots
  if (typeof sat === 'undefined' || sat.static) return;

  if (timeManager.realTime * 1 > settingsManager.lastBoxUpdateTime * 1 + updateInterval) {
    keepTrackApi.methods.updateSelectBox(sat);
    settingsManager.lastBoxUpdateTime = timeManager.realTime;
  }
};
