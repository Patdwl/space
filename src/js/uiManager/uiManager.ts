/**
 * /*! /////////////////////////////////////////////////////////////////////////////
 
 * http://keeptrack.space
 *
 * @Copyright (C) 2016-2022 Theodore Kruczek
 * @Copyright (C) 2020-2022 Heather Kruczek
 * @Copyright (C) 2015-2016, James Yoder
 *
 * Original source code released by James Yoder at https://github.com/jeyoder/ThingsInSpace/
 * under the MIT License. Please reference http://keeptrack.space/license/thingsinspace.txt
 *
 * KeepTrack is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * KeepTrack is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * KeepTrack. If not, see <http://www.gnu.org/licenses/>.
 *
 * /////////////////////////////////////////////////////////////////////////////
 */

// organize-imports-ignore
import 'jquery-ui-bundle';
import '@app/js/lib/external/jquery-ui-slideraccess.js';
import '@app/js/lib/external/jquery-ui-timepicker.js';
import '@app/js/lib/external/colorPick.js';
import '@materializecss/materialize';
// eslint-disable-next-line sort-imports
import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { clickAndDragHeight, closeColorbox, getEl } from '@app/js/lib/helpers';
import $ from 'jquery';
import { legendColorsChange, initUiValidation, uiInput, useCurrentGeolocationAsSensor, keyHandler, initMenuController } from '.';
import { searchToggle } from './search/searchToggle';
import { updateURL } from './url/update-url';
import { searchBox } from './search/searchBox';
import { doSearch } from './search/doSearch';
import { toast } from './ui/toast';
import { hideUi } from './ui/hideUi';
import { mobileManager } from './mobile/mobileManager';
import { hideLoadingScreen } from './ui/hideLoadingScreen';
import { colorSchemeChangeAlert } from './ui/colorSchemeChangeAlert';
import { legendHoverMenuClick, legendMenuChange } from './legend/legendMenu';
import { updateSelectBox } from './updateSelectBox';
import { getsensorinfo } from './ui/getsensorinfo';
import { footerToggle } from './ui/footerToggle';
import { loadStr } from './ui/loadStr';
import { reloadLastSensor } from './ui/reloadLastSensor';
import { panToStar } from './ui/panToStar';

// materializecss/materialize goes to window.M, but we want a local reference
export const M = window.M;

export let updateInterval = 1000;

export const init = () => {
  // Register all UI callbacks to run at the end of the draw loop
  keepTrackApi.register({
    method: 'onDrawLoopComplete',
    cbName: 'updateSelectBox',
    cb: updateSelectBox,
  });

  if (settingsManager.isShowLogo) getEl('demo-logo').classList.remove('start-hidden');

  keepTrackApi.methods.uiManagerInit();

  initBottomMenuResizing();

  // Initialize Navigation and Select Menus
  let elems;
  elems = document.querySelectorAll('.dropdown-button');
  M.Dropdown.init(elems);
};
export const initBottomMenuResizing = () => {
  // Allow Resizing the bottom menu
  const maxHeight = getEl('bottom-icons') !== null ? getEl('bottom-icons').offsetHeight : 0;
  clickAndDragHeight(getEl('bottom-icons-container'), maxHeight, () => {
    let bottomHeight = getEl('bottom-icons-container').offsetHeight;
    document.documentElement.style.setProperty('--bottom-menu-height', bottomHeight + 'px');
    if (window.getComputedStyle(getEl('nav-footer')).bottom !== '0px') {
      document.documentElement.style.setProperty('--bottom-menu-top', '0px');
    } else {
      bottomHeight = getEl('bottom-icons-container').offsetHeight;
      document.documentElement.style.setProperty('--bottom-menu-top', bottomHeight + 'px');
    }
  });
};
export const onReady = () => {
  // Code Once index.htm is loaded
  if (settingsManager.offline) updateInterval = 250;

  // Load Bottom icons
  if (!settingsManager.disableUI) {
    $.event.special.touchstart = {
      setup: function (_, ns, handle: any) {
        if (ns.includes('noPreventDefault')) {
          this.addEventListener('touchstart', handle, { passive: false });
        } else {
          this.addEventListener('touchstart', handle, { passive: true });
        }
      },
    };
  }

  (function _menuInit() {
    $('.tooltipped').tooltip(<any>{ delay: 50 });

    // Setup Legend Colors
    legendColorsChange();
  })();

  uiManager.clearRMBSubMenu = () => {
    getEl('save-rmb-menu').style.display = 'none';
    getEl('view-rmb-menu').style.display = 'none';
    getEl('edit-rmb-menu').style.display = 'none';
    getEl('create-rmb-menu').style.display = 'none';
    getEl('colors-rmb-menu').style.display = 'none';
    getEl('draw-rmb-menu').style.display = 'none';
    getEl('earth-rmb-menu').style.display = 'none';
  };

  uiManager.menuController = initMenuController;

  // Run any plugins code
  keepTrackApi.methods.uiManagerOnReady();
  uiManager.bottomIconPress = (el: HTMLElement) => keepTrackApi.methods.bottomMenuClick(el.id);
  getEl('bottom-icons').addEventListener('click', function (evt: Event) {
    if ((<HTMLElement>evt.target).id === 'bottom-icons') return;
    if ((<HTMLElement>evt.target).parentElement.id === 'bottom-icons') {
      uiManager.bottomIconPress(<HTMLElement>evt.target);
    } else {
      uiManager.bottomIconPress((<HTMLElement>evt.target).parentElement);
    }
  });
  uiManager.hideSideMenus = () => {
    closeColorbox();
    keepTrackApi.methods.hideSideMenus();
  };
  (<any>$('#bottom-icons')).sortable({ tolerance: 'pointer' });
};
// This runs after the drawManager starts
export const postStart = () => {
  initUiValidation();

  setTimeout(() => {
    document.querySelectorAll('img').forEach((img: any) => {
      if (!img.src.includes('.png') && !img.src.includes('.jpg')) {
        img.src = img.attributes.delayedsrc?.value;
      }
    });
  }, 0);

  (function _httpsCheck() {
    if (location.protocol !== 'https:') {
      try {
        getEl('cs-geolocation').style.display = 'none';
        getEl('geolocation-btn').style.display = 'none';
      } catch {
        // Intended to catch errors when the page is not loaded yet
      }
    }
  })();

  // Enable Satbox Overlay
  if (settingsManager.enableHoverOverlay) {
    try {
      const hoverboxDOM = document.createElement('div');
      hoverboxDOM.innerHTML = `
        <div id="sat-hoverbox">
          <span id="sat-hoverbox1"></span>
          <br/>
          <span id="sat-hoverbox2"></span>
          <br/>
          <span id="sat-hoverbox3"></span>
        </div>`;

      getEl('keeptrack-canvas').parentElement.append(hoverboxDOM);
    } catch {
      /* istanbul ignore next */
      console.debug('document.createElement() failed!');
    }
  }
};

export type UiManager = typeof uiManager;
export const uiManager = {
  lastBoxUpdateTime: 0,
  hideUi,
  legendColorsChange,
  legendMenuChange,
  isUiVisible: false,
  keyHandler,
  uiInput,
  useCurrentGeolocationAsSensor,
  searchBox,
  mobileManager,
  isCurrentlyTyping: false,
  onReady,
  init,
  postStart,
  getsensorinfo,
  footerToggle,
  searchToggle,
  hideSideMenus: null,
  hideLoadingScreen,
  loadStr,
  colorSchemeChangeAlert,
  lastColorScheme: null,
  updateURL,
  lookAtLatLon: null,
  reloadLastSensor,
  doSearch,
  panToStar,
  clearRMBSubMenu: null,
  menuController: null,
  legendHoverMenuClick,
  bottomIconPress: null,
  toast,
  createClockDOMOnce: false,
  lastNextPassCalcSatId: 0,
  lastNextPassCalcSensorId: null,
  resize2DMap: null,
  isAnalysisMenuOpen: false,
  updateNextPassOverlay: null,
  earthClicked: null,
};
