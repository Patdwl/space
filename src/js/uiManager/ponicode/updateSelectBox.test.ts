import { keepTrackApiStubs } from '@app/js/api/apiMocks';
import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { KeepTrackPrograms } from '@app/js/api/keepTrackTypes';
import * as updateSelectBox from '@app/js/uiManager/updateSelectBox';

keepTrackApi.programs = <KeepTrackPrograms>(<unknown>{ ...keepTrackApi.programs, ...keepTrackApiStubs.programs });

// @ponicode
describe('updateSelectBox.updateSelectBox', () => {
  test('0', () => {
    let result: any = updateSelectBox.updateSelectBox();
    expect(result).toMatchSnapshot();
  });
});
