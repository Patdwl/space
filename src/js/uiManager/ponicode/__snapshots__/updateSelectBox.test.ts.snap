// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateSelectBox.updateSelectBox 0 1`] = `undefined`;
